gufe/protocols/protocoldag.py

```diff
@@ -141,9 +141,15 @@ def result_to_unit(self, protocol_unit_result: ProtocolUnitResult):
             raise KeyError("No such `protocol_unit_result` present")
 
     def ok(self) -> bool:
-        # ensure that for every protocol unit, there is definitely a
-        # corresponding result
-        return set(self._protocol_units) == set(self._unit_result_mapping.keys())
+        # ensure there are no failed untis and that for every protocol unit,
+        # there is definitely a corresponding result
+        return (len(self.protocol_unit_failures) == 0 and
+                set(self._protocol_units) == set(self._unit_result_mapping.keys()))
+
+
+    # TODO: add a convenience method to get terminal protocol units
+    def terminal_protocol_unit_results(self):
+        ...
```

Comment:
I can't remember which way round our DAG is, but this is something like:
 `[pur for pur in self.protocol_unit_results if not nx.ancestors(self._result_graph, pur)]`

Reply:
Something like that! I think this will be useful for protocol authors in writing their `Protocol._gather` methods, since they'll likely pull things out of the `ProtocolUnitResult`s with no dependencies. I'm considering using terminal nodes as the basis for what paths get pushed to the objectstore in `fah-alchemy`, so probably would use this method in my compute services as well.